File: book/petrol-pump/java/chapter4/section8/ClearSalePump.java
package chapter4.section8;

import pump.*;
import chapter4.section4.LifeCycle;
import chapter4.section7.Fill;
import chapter4.section7.ShowDollarsPump;
import nz.sodium.*;
import java.util.Optional;

public class ClearSalePump implements Pump {
    public Outputs create(Inputs inputs) {
        StreamLoop<Fuel> sStart = new StreamLoop<>();
        Fill fi = new Fill(
                          inputs.sClearSale.map(u -> Unit.UNIT),
                          inputs.sFuelPulses, inputs.calibration,
                          inputs.price1, inputs.price2, inputs.price3,
                          sStart);
        NotifyPointOfSale np = new NotifyPointOfSale(
                new LifeCycle(inputs.sNozzle1,
                              inputs.sNozzle2,
                              inputs.sNozzle3),
                inputs.sClearSale,
                fi);
        sStart.loop(np.sStart);
        return new Outputs()
            .setDelivery(np.fuelFlowing.map(
                of ->
                    of.equals(Optional.of(Fuel.ONE))   ? Delivery.FAST1 :
                    of.equals(Optional.of(Fuel.TWO))   ? Delivery.FAST2 :
                    of.equals(Optional.of(Fuel.THREE)) ? Delivery.FAST3 : 
                                                         Delivery.OFF))
            .setSaleCostLCD(fi.dollarsDelivered.map(
                    q -> Formatters.formatSaleCost(q)))
            .setSaleQuantityLCD(fi.litersDelivered.map(
                    q -> Formatters.formatSaleQuantity(q)))
            .setPriceLCD1(ShowDollarsPump.priceLCD(np.fillActive, fi.price,
                    Fuel.ONE, inputs))
            .setPriceLCD2(ShowDollarsPump.priceLCD(np.fillActive, fi.price,
                    Fuel.TWO, inputs))
            .setPriceLCD3(ShowDollarsPump.priceLCD(np.fillActive, fi.price,
                    Fuel.THREE, inputs))
            .setBeep(np.sBeep)
            .setSaleComplete(np.sSaleComplete);
    }
}

File: book/petrol-pump/c#/PetrolPump/Chapter4/Section8/ClearSalePump.cs
using PetrolPump.Chapter4.Section4;
using PetrolPump.Chapter4.Section7;
using Sodium.Frp;
using Sodium.Functional;

namespace PetrolPump.Chapter4.Section8
{
    public class ClearSalePump : IPump
    {
        public Outputs Create(Inputs inputs)
        {
            StreamLoop<Fuel> sStart = new StreamLoop<Fuel>();
            Fill fi = new Fill(
                inputs.SClearSale,
                inputs.SFuelPulses, inputs.Calibration,
                inputs.Price1, inputs.Price2, inputs.Price3,
                sStart);
            NotifyPointOfSale np = new NotifyPointOfSale(
                new LifeCycle(inputs.SNozzle1,
                    inputs.SNozzle2,
                    inputs.SNozzle3),
                inputs.SClearSale,
                fi);
            sStart.Loop(np.SStart);
            return new Outputs()
                .SetDelivery(np.FuelFlowing.Map(
                    m =>
                        m.Equals(Maybe.Some(Fuel.One)) ? Delivery.Fast1 :
                            m.Equals(Maybe.Some(Fuel.Two)) ? Delivery.Fast2 :
                                m.Equals(Maybe.Some(Fuel.Three)) ? Delivery.Fast3 :
                                    Delivery.Off))
                .SetSaleCostLcd(fi.DollarsDelivered.Map(Formatters.FormatSaleCost))
                .SetSaleQuantityLcd(fi.LitersDelivered.Map(Formatters.FormatSaleQuantity))
                .SetPriceLcd1(ShowDollarsPump.PriceLcd(np.FillActive, fi.Price, Fuel.One, inputs))
                .SetPriceLcd2(ShowDollarsPump.PriceLcd(np.FillActive, fi.Price, Fuel.Two, inputs))
                .SetPriceLcd3(ShowDollarsPump.PriceLcd(np.FillActive, fi.Price, Fuel.Three, inputs))
                .SetBeep(np.SBeep)
                .SetSaleComplete(np.SSaleComplete);
        }
    }
}